hikari/events/channel_events.py

```diff
@@ -780,3 +788,264 @@ async def fetch_guild_webhooks(self) -> typing.Sequence[webhooks.PartialWebhook]
             If an internal error occurs on Discord while handling the request.
         """
         return await self.app.rest.fetch_guild_webhooks(self.guild_id)
+
+
+@base_events.requires_intents(intents.Intents.GUILDS, intents.Intents.GUILDS | intents.Intents.GUILD_MEMBERS)
+class GuildThreadEvent(shard_events.ShardEvent, abc.ABC):
+    """Event base for any event that is related to a guild thread."""
+
+    __slots__: typing.Sequence[str] = ()
+
+    @property
+    @abc.abstractmethod
+    def guild_id(self) -> snowflakes.Snowflake:
+        """ID of the guild this event is for."""
+
+    @property
+    @abc.abstractmethod
+    def thread_id(self) -> snowflakes.Snowflake:
+        """ID of the thread this event is for."""
+
+    async def fetch_channel(self) -> channels.GuildThreadChannel:
+        """Perform an API call to fetch the details about this thread.
+
+        !!! note
+            For `GuildThreadDeleteEvent` events, this will always raise
+            an exception, since the channel will have already been removed.
+
+        Returns
+        -------
+        hikari.channels.GuildThreadChannel
+            A derivative of `hikari.channels.GuildThreadChannel`. The
+            actual type will vary depending on the type of channel this event
+            concerns.
+
+        Raises
+        ------
+        hikari.errors.UnauthorizedError
+            If you are unauthorized to make the request (invalid/missing token).
+        hikari.errors.ForbiddenError
+            If you are missing the `READ_MESSAGES` permission in the channel.
+        hikari.errors.NotFoundError
+            If the channel is not found.
+        hikari.errors.RateLimitTooLongError
+            Raised in the event that a rate limit occurs that is
+            longer than `max_rate_limit` when making a request.
+        hikari.errors.RateLimitedError
+            Usually, Hikari will handle and retry on hitting
+            rate-limits automatically. This includes most bucket-specific
+            rate-limits and global rate-limits. In some rare edge cases,
+            however, Discord implements other undocumented rules for
+            rate-limiting, such as limits per attribute. These cannot be
+            detected or handled normally by Hikari due to their undocumented
+            nature, and will trigger this exception if they occur.
+        hikari.errors.InternalServerError
+            If an internal error occurs on Discord while handling the request.
+        """
+        channel = await self.app.rest.fetch_channel(self.thread_id)
+        assert isinstance(channel, channels.GuildThreadChannel)
+        return channel
+
+
+@base_events.requires_intents(intents.Intents.GUILDS)
+@attr_extensions.with_copy
+@attr.define(kw_only=True, weakref_slot=False)
+class GuildThreadAccessEvent(GuildThreadEvent):
```

Comment:
What would the event be for when you are removed access?

Reply:
if you were in the thread and removed from the thread itself then THREAD_MEMBERS_UPDATE, but there's no dedicated event for when you lose the perms to "access" a thread that and you'll have to check based on member updates, channel updates, guild updates, guild remove, channel deletes,